Implement `wolfSSL_set_client_CA_list` and add 'HIGH' cipher suite

julek-wolfssl opened the pull request.

julek-wolfssl committed:
Implement `wolfSSL_set_client_CA_list` and add 'HIGH' cipher suite

julek-wolfssl committed:
Parse distinguished names in `DoCertificateRequest`

The CA names sent by the server are now being parsed in `DoCertificateRequest` and are saved on a stack in `ssl->ca_names`.

julek-wolfssl commented:
@haydenroche5 Could you take a look if the macro guard change is acceptable in `ssl.c`? I'm referring to the part with the comment `Note: This is a huge section of API's`.

haydenroche5 commented on src/internal.c:
```diff
@@ -2194,14 +2194,10 @@ void SSL_CtxResourceFree(WOLFSSL_CTX* ctx)
         }
     #endif
-    #ifdef OPENSSL_EXTRA
+    #if defined(OPENSSL_EXTRA) || defined(WOLFSSL_EXTRA) || defined(HAVE_LIGHTY)
```
What is `WOLFSSL_EXTRA`? I can't find it in configure.ac. Looks like it's only used to gate in compilation of `ca_names` in internal.h. Would be good to see if this macro can be removed.

haydenroche5 replied:
Still wondering about this.

julek-wolfssl replied:
Oh, I'm not sure where `WOLFSSL_EXTRA` is coming from. It's how `ca_names` is defined in the `WOLFSSL_CTX` struct so it should be `free`'d in the same conditions. I'll ask if anyone knows anything about this define.

haydenroche5 replied:
For future readers: 8d0731a5d83e902d1041b521d776a3f90fec844c

haydenroche5 replied:
This looks like the right macro guard for `ctx->ca_names`, but not `ctx->x509_store_pt`. `ctx->x509_store_pt` exists only if `#if defined(OPENSSL_EXTRA) || defined(HAVE_WEBSERVER) || defined(WOLFSSL_WPAS_SMALL)`.

julek-wolfssl replied:
Changing guard.

haydenroche5 commented on src/internal.c:
```diff
@@ -10296,37 +10320,22 @@ int CopyDecodedToX509(WOLFSSL_X509* x509, DecodedCert* dCert)
-    XSTRNCPY(x509->issuer.name, dCert->issuer, ASN_NAME_MAX);
-    x509->issuer.name[ASN_NAME_MAX - 1] = '\0';
-    x509->issuer.sz = (int)XSTRLEN(x509->issuer.name) + 1;
+    CopyDecodedName(&x509->issuer, dCert, ISSUER);
```
Nice refactor.

haydenroche5 commented on src/ssl.c:
```diff
@@ -43358,10 +43395,7 @@ void* wolfSSL_SESSION_get_ex_data(const WOLFSSL_SESSION* session, int idx)
-    (defined(HAVE_STUNNEL) || defined(WOLFSSL_NGINX) || \
-    defined(HAVE_LIGHTY) || defined(WOLFSSL_HAPROXY) || \
-    defined(WOLFSSL_OPENSSH) || defined(HAVE_SBLIM_SFCB)))
+#if defined(OPENSSL_EXTRA)
```
I wouldn't do this, because it will increase the code size for people using `OPENSSL_EXTRA` without `OPENSSL_ALL` and without one of `HAVE_STUNNEL`, `WOLFSSL_NGINX`, etc.

julek-wolfssl replied:
I understand but what is the logic of adding this section only when `OPENSSL_EXTRA` is used with one of the project defines (many of which define `OPENSSL_EXTRA` anyway). I need some things from this section for the `ca_names` handling. Alternatively, I can hide that stuff behind an `OPENSSL_ALL` guard.

haydenroche5 replied:
My guess at the logic would be that we don't want the code included for users just using `OPENSSL_EXTRA`, without any of  the items from the list of projects (OpenSSH, stunnel, etc.). The goal being reducing code size for that use case. Is it an important use case? I have no idea, so maybe the change is fine. @cconlon do you have any thoughts?

julek-wolfssl replied:
I changed this back to how it was. I moved the `wolfSSL_sk_X509_NAME_*` functions outside of this boundary into an `OPENSSL_EXTRA` block. These are needed for the `client_ca` functions.

julek-wolfssl replied:
Without these functions we are re-implementing a lot of the stack manipulation code.

haydenroche5 reviewed (changes requested):
I think the macro guard around the big section of APIs should be left as-is.

julek-wolfssl committed:
Reorganize wolfSSL_sk_X509_NAME_*

Make the `wolfSSL_sk_X509_NAME_*` API's available in OPENSSL_EXTRA for use with `client_CA_list` API's.

julek-wolfssl committed:
Refactor `client_CA` API to use `wolfSSL_sk_X509_NAME_*` API

julek-wolfssl commented:
Retest this please.

haydenroche5 commented on src/ssl.c:
```diff
@@ -15902,18 +15902,48 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
+                                       WOLF_STACK_OF(WOLFSSL_X509_NAME)* names)
+    {
+        WOLFSSL_ENTER("wolfSSL_set_client_CA_list");
+    #if defined(OPENSSL_EXTRA) || defined(WOLFSSL_EXTRA)
```
Why is this same macro guard not needed for `wolfSSL_CTX_get_client_CA_list`?

julek-wolfssl replied:
Not needed indeed. The functions are now protected by this guard. Removing redundant.

haydenroche5 commented on src/ssl.c:
```diff
@@ -15924,198 +15954,150 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
-        if (s == NULL)
-            return NULL;
+        if (wolfSSL_sk_X509_NAME_push(ctx->ca_names, nameCopy) != WOLFSSL_SUCCESS) {
+            WOLFSSL_MSG("wolfSSL_sk_X509_NAME_push error");
```
In this error case, should `ctx->ca_names` be freed if it was allocated in this function call? I.e. if the conditional at line 15972 was taken.

julek-wolfssl replied:
`ctx->ca_names` will be `free`'d when `free`'ing the entire `WOLFSSL_CTX` structure. We don't need to do it here.

haydenroche5 commented on src/ssl.c:
```diff
@@ -43856,88 +43843,286 @@ void* wolfSSL_SESSION_get_ex_data(const WOLFSSL_SESSION* session, int idx)
-    defined(WOLFSSL_OPENSSH) || defined(HAVE_SBLIM_SFCB)))
-int wolfSSL_SESSION_get_ex_new_index(long idx, void* data, void* cb1,
-       void* cb2, CRYPTO_free_func* cb3)
+#if defined(OPENSSL_EXTRA) || defined(HAVE_STUNNEL) || defined(WOLFSSL_NGINX) || \
```
I still don't really understand why this change is being made.

> Without these functions we are re-implementing a lot of the stack manipulation code.

The functions only exist once in ssl.c; what do you mean by this? I don't see how we're re-implementing anything.

julek-wolfssl replied:
If you look in `internal.c` and `ssl.c` for this PR, you will find many instances where I refactored functions to use the `wolfSSL_sk_X509_NAME_*` API instead of directly manipulating the stack. Examples include `SSL_CtxResourceFree`, `wolfSSL_load_client_CA_file`, `wolfSSL_CTX_add_client_CA`. I believe that's all of them in this PR but there may be more.

haydenroche5 commented on tests/api.c:
```diff
@@ -30556,26 +30556,118 @@ static void test_wolfSSL_CTX_set_client_CA_list(void)
     printf(testingFmt, "wolfSSL_CTX_set_client_CA_list()");
-    AssertNotNull(ctx = wolfSSL_CTX_new(wolfSSLv23_client_method()));
+    AssertNotNull(ctx = wolfSSL_CTX_new(wolfSSLv23_server_method()));
+    /* Send two X501 names in cert request */
```
I think x501 here should be x509.

julek-wolfssl replied:
https://datatracker.ietf.org/doc/html/rfc5246#section-7.4.4 cites a x501 spec for the distinguished name. I've found Red Hat referring to it as x500 names. I'll change it to whatever you think is best.

haydenroche5 replied:
Works for me, I just thought it was a typo.

haydenroche5 commented on src/internal.c:
```diff
@@ -10278,6 +10278,30 @@ static void AddSessionCertToChain(WOLFSSL_X509_CHAIN* chain,
+        XSTRNCPY(name->name, dCert->subject, ASN_NAME_MAX);
+        name->name[ASN_NAME_MAX - 1] = '\0';
+        name->sz = (int)XSTRLEN(name->name) + 1;
+#if defined(OPENSSL_ALL) || defined(WOLFSSL_NGINX)
```
Looks like `name->raw` and `name->rawLen` exist if `#if defined(OPENSSL_ALL) || defined(WOLFSSL_NGINX) || defined(HAVE_LIGHTY)`. Macro guard here should match.

julek-wolfssl replied:
Fixed to match.

haydenroche5 commented on src/internal.c:
```diff
@@ -10278,6 +10278,30 @@ static void AddSessionCertToChain(WOLFSSL_X509_CHAIN* chain,
+        XSTRNCPY(name->name, dCert->issuer, ASN_NAME_MAX);
+        name->name[ASN_NAME_MAX - 1] = '\0';
+        name->sz = (int)XSTRLEN(name->name) + 1;
+#if (defined(OPENSSL_ALL) || defined(WOLFSSL_NGINX)) && defined(WOLFSSL_CERT_EXT)
```
I think this should be #if (defined(OPENSSL_ALL) || defined(WOLFSSL_NGINX) || defined(HAVE_LIGHTY)) && (defined(WOLFSSL_CERT_EXT) || defined(HAVE_PKCS7)).

julek-wolfssl replied:
Fixed.

haydenroche5 commented on src/internal.c:
```diff
@@ -22743,6 +22765,37 @@ int DecodePrivateKey(WOLFSSL *ssl, word16* length)
+
+                if (wolfSSL_sk_X509_NAME_push(ssl->ca_names, name)
+                        == WOLFSSL_FAILURE) {
+                    FreeDecodedCert(&cert);
```
Should `name` also be freed in this error case?

julek-wolfssl replied:
Yup.

haydenroche5 commented on src/ssl.c:
```diff
@@ -15924,198 +15943,150 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
-    /* registers client cert callback, called during handshake if server
-       requests client auth but user has not loaded client cert/key */
-    void wolfSSL_CTX_set_client_cert_cb(WOLFSSL_CTX *ctx, client_cert_cb cb)
+    #if !defined(NO_RSA) && !defined(NO_CERTS)
```
It seems weird to me that this requires `!defined(NO_RSA)`. Are we sure that's necessary?

julek-wolfssl replied:
Removed. Might not be necessary.

haydenroche5 commented on src/ssl.c:
```diff
@@ -15924,198 +15943,150 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
-        WOLFSSL_ENTER("wolfSSL_CTX_get_client_CA_list");
+        nameCopy = wolfSSL_X509_NAME_dup(wolfSSL_X509_get_subject_name(x509));
+        if (nameCopy == NULL) {
+            WOLFSSL_MSG("wolfSSL_X509_NAME_dup error");
```
If `ctx->ca_names` was allocated in this function, should it be freed here?

haydenroche5 replied:
Oh, you may be able to disregard where I've made this comment. I forgot you said it'll be freed when the CTX is ultimately freed.

haydenroche5 commented on src/ssl.c:
```diff
@@ -15924,198 +15943,150 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
-        if (s == NULL)
-            return NULL;
+        if (wolfSSL_sk_X509_NAME_push(ctx->ca_names, nameCopy) != WOLFSSL_SUCCESS) {
+            WOLFSSL_MSG("wolfSSL_sk_X509_NAME_push error");
```
Same here.

haydenroche5 commented on src/ssl.c:
```diff
@@ -15924,198 +15943,150 @@ int wolfSSL_set_compression(WOLFSSL* ssl)
         }
+        #endif
+    #endif /* !NO_BIO */
+#endif /* OPENSSL_EXTRA || WOLFSSL_EXTRA || HAVE_WEBSERVER */
```
Comment here is wrong. `HAVE_WEBSERVER` isn't part of the macro guard.

julek-wolfssl replied:
Fixed.

haydenroche5 commented on src/ssl.c:
```diff
@@ -18800,7 +18773,10 @@ WOLF_STACK_OF(WOLFSSL_X509)* wolfSSL_set_peer_cert_chain(WOLFSSL* ssl)
-    sk = wolfSSL_sk_X509_new();
+    if (ssl->peerCertChain == NULL)
+        sk = wolfSSL_sk_X509_new();
+    else /* Try to re-use old chain if available */
```
If reusing, should the stack be emptied before proceeding? What if there are items in the stack?

julek-wolfssl replied:
You're right. This doesn't make sense. I'll just free the existing struct if it exists.

haydenroche5 commented on wolfcrypt/src/asn.c:
```diff
@@ -6535,9 +6536,17 @@ int GetName(DecodedCert* cert, int nameType, int maxIdx)
 #if (defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)) && \
             !defined(WOLFCRYPT_ONLY)
     if (nameType == ISSUER) {
+#if (defined(OPENSSL_ALL) || defined(WOLFSSL_NGINX)) && defined(WOLFSSL_CERT_EXT)
```
I think this macro guard and the one below might not be totally right. See my previous comment where `HAVE_PKCS7` needed to be checked.

julek-wolfssl replied:
Fixed.

haydenroche5 replied:
I don't see `HAVE_PKCS7` here. Perhaps the fix was left out of the latest push by mistake?

julek-wolfssl replied:
Yup. Put it back in.

haydenroche5 reviewed (changes requested):
No glaring problems, just some minor comments.

julek-wolfssl committed:
Code review fixes.

dgarske commented on wolfcrypt/src/asn.c:
```diff
@@ -136,6 +136,7 @@ extern int wc_InitRsaHw(RsaKey* key);
 #endif
 
 #if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+    #include <wolfssl/internal.h>
```
Can we avoid including `wolfssl/internal.h`? This seems problematic going forward and breaks the wolfssl/wolfcrypt separation. If there are things asn.c needs they should be in asn.h.

julek-wolfssl replied:
`wolfssl/openssl/objects.h` is already included on the next line which in turn includes `wolfssl/ssl.h`. Removing the second include results in a long list of missing symbols meaning that `asn.c` is already very closely tied to wolfSSL. I don't think adding this include changes much in the wolfssl/wolfcrypt separation. Someone could in the future investigate the feasibility to refactor out the wolfssl parts back into the `src` folder.

dgarske reviewed (changes requested):
Re-org is making ssl.c review difficult. Can you do the re-org as a different PR later? Also consider squash merge. The "HIGH" logic just returns successfully and lets wolfSSL default be used. Is that sufficient?

julek-wolfssl commented:
@dgarske The re-org of the API's in `ssl.c` is quite essential in this PR. The API's moved pertain to handling the `X509_NAME` stack. The `CA_list` is a stack of these names so the API is needed. I don't think doing this as a separate PR will be much easier as  I would have to manually manipulate the stack in the `CA_list` APIs for this PR.

dgarske commented:
@julek-wolfssl please resolve merge conflicts. Thanks

julek-wolfssl committed:
Rebase fixes

julek-wolfssl commented:
@dgarske Conflicts resolved. Waiting for Jenkins.
